### PR TITLE
Allow WooPay to request session data from merchant

### PR DESCRIPTION
**Pull request by bborman22**
#### Changes proposed in this Pull Request

This PR is to address a performance optimization in the WooPay Direct Checkout feature. This PR is one half of the fix for https://github.com/Automattic/woopay/issues/2480. This PR should be tested in tandem with https://github.com/Automattic/woopay/pull/2529.

The overall feature is to send a minimum amount of session data to WooPay to start, then let WooPay request the rest of the session data only if it is needed (i.e. the shopper is already logged into WooPay). If the shopper is not logged into WooPay the full session data should not be requested and the shopper should be taken to the merchant checkout page. More details about the feature and this optimization can be found in https://github.com/Automattic/woopay/issues/2480.

To accomplish that goal, this PR adds the following:
- The ability to retrieve the minimum session data either on page load or through AJAX if necessary.
- ~~Pass the minimum session data to the connect page and receive a redirect URL.~~
- Pass the encrypted and signed minimum session data to WooPay with a redirect.
- A new REST endpoint that allows WooPay to request the full session data from the merchant.

The points of failure I focused on in this PR:
- Inability to get the minimum session data.
- Inability to retrieve full session data from the merchant.

In all cases when there is a failure there, we should default to taking the shopper to the merchant checkout page.

**Notes**
One key thing to note is the need to unset the `preloaded_requests` on the full session data response. This is unfortunate as it reduces some of the performance benefits. The reason this had to be unset is that the preloaded requests use `rest_preload_api_request` which just replicates a REST request but with no headers, which means no `Cart-Token`. This leads the cart not being found and instead returning an empty cart object breaking things on WooPay. I experimented with other ways to request the cart data doing something like this:

```
add_filter( 'woocommerce_store_api_disable_nonce_check', '__return_true' );
$cart_request = new WP_REST_Request('GET', '/wc/store/v1/cart' );
$cart_request->set_headers( $woopay_request->get_headers() );
$cart_data = rest_do_request( $cart_request );

$checkout_request = new WP_REST_Request('GET', '/wc/store/v1/checkout' );
$checkout_request->set_headers( $woopay_request->get_headers() );
$checkout_data = rest_do_request( $checkout_request );
remove_filter( 'woocommerce_store_api_disable_nonce_check', '__return_true' );
```

But ran into the problem that the initial requests had been fired and so WooCommerce wasn't trying to load the cart again since it already loaded it on the first requests in `get_init_session_request` and it generally seemed "too late" to update the current session to use the Cart Token. **Given the time constraints I opted to leave the unset in for now and revisit in a separate issue, but if we find a good solution I'd love to add it to this PR!**

#### Testing instructions

All tests should start with WooPay enabled, the `_wcpay_feature_woopay_direct_checkout` option set to 1 (enabled), and an item in the cart.

- When the shopper is logged out of WooPay, clicking the `Proceed to Checkout` button should take the shopper to the merchant checkout. If third party cookies are disabled, there will be a request to `woopay/?checkout_redirect=1&blog_id=...` that results in a 302 redirect to the merchant checkout. In this case nothing should ever get stored on WooPay for the session.
- When the shopper is logged in to WooPay,  clicking the `Proceed to Checkout` button should take the shopper to the WooPay checkout. If third party cookies are disabled, there will be a request to `woopay/?checkout_redirect=1&blog_id=...` that results in a 302 redirect to WooPay with a `platform_checkout_key`. 
- Test the existing flows are unaffected:
    - Using the WooPay Express checkout button on the cart and checkout page.
    - Using the OTP modal on checkout page after entering an email address.
    - Checking out without WooPay.
- Test in both blocks and shortcode cart.
- Test with a shopper that is logged into the merchant store.
- Test with the WooPay PR running on the staging sandbox to confirm environment compatibility. 

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8268` or branch name `add/woopay-merchant-request` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=add/woopay-merchant-request) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: a265f797ff52cb9d96c8793262f3690b97e4f7f7
- Build time: 2024-03-21 22:44:09 UTC

_Note: the build is updated when a new commit is pushed to this PR._

**Comment by bborman22** [comment]
Adding @ricardo as a reviewer given the round robin assignment from the WooPay PR.

File: client/checkout/woopay/direct-checkout/woopay-direct-checkout.js [hunk 1]
```diff
@@ -234,6 +279,51 @@ class WooPayDirectCheckout {
...
 
+	/**
+	 * Gets the WooPay redirect data.
+	 *
+	 * @return {Promise<Promise<*>|*>} Resolves to the WooPay redirect response.
+	 */
+	static async getWooPayMinimumSesssionDataFromMerchant() {
+		if ( getConfig( 'woopayMinimumSessionData' ) ) {
+			return getConfig( 'woopayMinimumSessionData' );
+		}
```
Comment (ricardo) [inline client/checkout/woopay/direct-checkout/woopay-direct-checkout.js:357]: Shouldn't this always be defined from the cart page? Or are we making the request below just in case the user clicks the checkout link from a different page when this feature becomes available?

Just to understand if I'm missing something.

It would be nice to add some comment explaining when we expect this to not be available.

File: tests/e2e/specs/wcpay/merchant/merchant-orders-full-refund.spec.js [hunk 2]
```diff
@@ -33,6 +33,15 @@ describe( 'Order > Full refund', () => {
...
 		await expect( page ).toMatch( 'Order received' );
 
+		// Get the order price
+		const priceElement = await page.$(
+			'.woocommerce-order-overview__total .woocommerce-Price-amount'
+		);
+		orderAmount = await page.evaluate(
+			( el ) => el.textContent,
+			priceElement
+		);
```
Comment (ricardo) [inline tests/e2e/specs/wcpay/merchant/merchant-orders-full-refund.spec.js]: I assume this was an unrelated fix that we're including just because it's small? 🙂

File: includes/admin/class-wc-rest-woopay-session-controller.php [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+ * Class WC_REST_WooPay_Adapted_Extensions_Controller
+ *
+ * @package WooCommerce\Payments\Admin
+ *
+ *
+ * NOTE: This API endpoint is never registered and not available for use.
+ * It is only available here as a temporary workaround for the issue described in
+ * https://github.com/Automattic/woocommerce-payments/issues/6304
+ *
+ * This file can be removed once the issue is resolved.
```
Comment (ricardo) [inline includes/admin/class-wc-rest-woopay-session-controller.php]: - This endpoint is in fact registered, right?
- Line 3 should read `WC_REST_WooPay_Session_Controller`.

File: includes/admin/class-wc-rest-woopay-session-controller.php [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+	 *
+	 * @var string
+	 */
+	protected $rest_base = 'woopay/session';
+
+	/**
+	 * Configure REST API routes.
+	 */
+	public function register_routes() {
+		// TODO: We need to give this proper permission checks, required fields, etc.
```
Comment (ricardo) [inline includes/admin/class-wc-rest-woopay-session-controller.php]: Do you have anything else in mind for the permission check other than `has_valid_request_signature`?

It might be a good idea to add the required fields 👍.

**Review by ricardo — CHANGES_REQUESTED** [review]
Dropped the main comment in the WooPay repo, but the approach looks good to me. Nice work!

Only left pretty minor questions on this PR.

All tests have passed:

✅ With third party cookies disabled, there's a request to woopay/?cache_checkout_key that redirects to the merchant checkout if the user is logged out.
✅ With third party cookies disabled, there's a request to woopay/?cache_checkout_key that redirects to the WooPay checkout if the user is logged in.
✅ Existing flows work as expected.
✅ Tested both in blocks and shortcode cart.
✅ Tested with a user logged in to the merchant store.
✅ Tested in sandbox with Jetpack blog auth.

File: client/checkout/woopay/direct-checkout/woopay-direct-checkout.js (hunk 1 above)
Comment (bborman22) [reply client/checkout/woopay/direct-checkout/woopay-direct-checkout.js:357]: Yeah, it should always be defined. I was just being extra cautious here with a request fallback in case of some edge cases. I've added a comment in https://github.com/Automattic/woocommerce-payments/pull/8268/commits/416f4d9c49cfe93f2999ab0f71becf4e3298862a to clarify that

File: tests/e2e/specs/wcpay/merchant/merchant-orders-full-refund.spec.js (hunk 2 above)
Comment (bborman22) [reply tests/e2e/specs/wcpay/merchant/merchant-orders-full-refund.spec.js]: Doh! Totally unreletad change I was using to get some local E2E tests to pass. I've removed this change in https://github.com/Automattic/woocommerce-payments/pull/8268/commits/dfc8581500df294c4d98f456682c9b341f5e9866 as it has nothing to do with the PR, sorry about that!

File: includes/admin/class-wc-rest-woopay-session-controller.php (hunk 3 above)
Comment (bborman22) [reply includes/admin/class-wc-rest-woopay-session-controller.php]: This was a copy/paste issue and updated it in https://github.com/Automattic/woocommerce-payments/pull/8268/commits/da4c85f5f2deabca9b4fb2c2fed2a7478529ca64.

File: includes/admin/class-wc-rest-woopay-session-controller.php (hunk 4 above)
Comment (bborman22) [reply includes/admin/class-wc-rest-woopay-session-controller.php]: That todo was outdated at this point. There aren't really any required fields on this request (at this point), though it relies on some headers. Those headers are then checked in the `is_request_from_woopay` and `has_valid_request_signature` functions so the fix here was to just remove the todo, but let me know if you see another opportunity to check these permissions.

**Review by ricardo — APPROVED** [review]
LGTM and tests well.

**Comment by bborman22** [comment]
@ricardo I have now updated the implementation to use an encrypted and signed session data passed through the redirect URL to WooPay instead of storing a temporary session on WooPay. This was based on my discussion with Systems and their recommendation that the storage could be unreliable in this use case. I have updated https://github.com/Automattic/woopay/pull/2529 as well to handle this implementation and so they should be tested again together. 

Sorry for the back and forth on this PR, but this should be the last reworking of the implementation now that we've had discussions more broadly.

File: client/checkout/woopay/direct-checkout/woopay-direct-checkout.js [hunk 5]
```diff
@@ -291,6 +345,68 @@ class WooPayDirectCheckout {
...
+	 * @return {boolean} True if all required fields are present, false otherwise.
+	 */
+	static validateEncryptedSessionData( encryptedSessionData ) {
+		return (
+			encryptedSessionData?.blog_id &&
+			encryptedSessionData?.data?.session &&
+			encryptedSessionData?.data?.iv &&
+			encryptedSessionData?.data?.hash
+		);
+	}
```
Comment (ricardo) [inline client/checkout/woopay/direct-checkout/woopay-direct-checkout.js]: There's a very similar method in this file – `isValidEncryptedSessionData`.

Interesting how we sometimes end up writing almost the same code in separate PRs without knowing. 😅

File: includes/woopay/class-woopay-session.php [hunk 6]
```diff
@@ -556,6 +532,54 @@ public static function ajax_get_woopay_session() {
...
+		$data = [
+			'blog_id'           => $blog_id,
+			'blog_rest_url'     => get_rest_url(),
+			'blog_checkout_url' => wc_get_checkout_url(),
+			'session_nonce'     => self::create_woopay_nonce( get_current_user_id() ),
+			'store_api_token'   => self::init_store_api_token(),
+		];
+
+		return WooPay_Utilities::encrypt_and_sign_data( $data );
+	}
```
Comment (ricardo) [inline includes/woopay/class-woopay-session.php]: To prevent issues related to `openssl` not being enabled on the host, I guess we should add a `WC_Payments_Features::is_client_secret_encryption_eligible()` check before `encrypt_and_sign_data`, similar to how it's done in `get_frontend_init_session_request`.

**Review by ricardo — CHANGES_REQUESTED** [review]
Thank you for all the thought and effort on this new approach. It indeed looks better than the previous cache idea. 🚀 

I left 2 minor comments, but overall LGTM and tests well! 🎉

File: client/checkout/woopay/direct-checkout/woopay-direct-checkout.js (hunk 5 above)
Comment (bborman22) [reply client/checkout/woopay/direct-checkout/woopay-direct-checkout.js]: 😆 Wow, can't believe I missed that! Now using just the existing function and removed the duplicate in https://github.com/Automattic/woocommerce-payments/pull/8268/commits/4b1b7d4c320e178232d3c1ebea1cfed829d7c355

File: includes/woopay/class-woopay-session.php (hunk 6 above)
Comment (bborman22) [reply includes/woopay/class-woopay-session.php]: Great catch, I've copied the same check here now in https://github.com/Automattic/woocommerce-payments/pull/8268/commits/4b1b7d4c320e178232d3c1ebea1cfed829d7c355.

**Review by ricardo — APPROVED** [review]
LGTM!
